### ISSUE TITLE
Validation does not work correctly on Material-ui input with key prop

### ISSUE THREAD

**Issue by ASHFAQPATWARI** [bluebill1049/react-hook-form#177]
**Describe the bug**
I am using material-ui input element. And application has multiple language. I change label prop on language change to pass new label. And i use key prop to trigger styles calculation. But after this, the form validations does not work correctly

**Expected behavior**
Validations should work fine.

**Screenshots**
Initial Render: No error and submit works fine
<img width="490" alt="Screen Shot 2019-07-31 at 4 23 44 PM" src="https://user-images.githubusercontent.com/5632217/62215448-a5927b80-b3af-11e9-912e-9a7ccc6e550f.png">

After language switch: Error does not go away

<img width="485" alt="Screen Shot 2019-07-31 at 4 24 05 PM" src="https://user-images.githubusercontent.com/5632217/62215449-a6c3a880-b3af-11e9-8fff-3d07b0ffce91.png">

**Desktop (please complete the following information):**
 - OS: ios
 - Browser: Chrome

**Additional context**
This is how my textfield from material-ui looks:

`
<TextField
                      name="email"
                      key={state.lang}
                      label={${intl.formatMessage({ id: 'generic.email' })} *}
                      margin="normal"
                      variant="outlined"
                      helperText={errors.email ? intl.formatMessage({ id: errors.email.message }) : ''}
                      inputRef={register}
                      error={errors.email ? true : false}
                    />`

**Comment by bluebill1049** [comment]
any chance to provide me with a codesandbox?

**Comment by ASHFAQPATWARI** [comment]
@bluebill1049 By language switch I mean, I am using react-intl to switch language of my application from English to Arabic. 
I will not be doing `reset` because the form is not exactly resetted, just the language is switched. 
Also, the form works fine if i don't use the `key` prop. I will try to prepare a codesandbox, do you have a sandbox template link ?

**Comment by bluebill1049** [comment]
I don't have a template but appreciate for prepare codesandbox. thanks, I will look into once u supply that 👍

**Comment by ASHFAQPATWARI** [comment]
@bluebill1049 here is the sandbox: https://codesandbox.io/s/react-hook-form-issue-on-first-click-gqcif

Steps to reproduce:
1. Click on change language button without changing anything.
2. Click on submit button and error will come
3. Click submit again and error will be gone

For the demo purpose, i have used defaultValue, however the same thing happens with browser autofill. When application is loaded, value are autofilled by browser. Then user switch the language and clicks submitted but gets the error

**Comment by bluebill1049** [comment]
Thanks! gonna take a look at it.

**Comment by bluebill1049** [comment]
ha, I find the issue. you didn't give the button a `type="button"`.

button without a type will automatically be assigned as `submit` 👍

**Comment by ASHFAQPATWARI** [comment]
@bluebill1049 which button are you talking about? If you are talking about the `Change Language` button then thats outside the form.

**Comment by bluebill1049** [comment]
Did you update the codesandbox? I think It was inside before...

**Comment by bluebill1049** [comment]
looking the issue now.

**Comment by bluebill1049** [comment]
what's this key doing?

**Comment by ASHFAQPATWARI** [comment]
@bluebill1049 the `key` prop is used by material-ui to recalculate the inline styles as the label width change because the width of text is variable. There is nothing unusual with this as this is the recommended way by material-ui.

If I remove the key prop, then styling of my component will not work correctly.

**Comment by stramel** [comment]
I don't believe the `key` prop is the issue here. I'm able to reproduce it with the codesandbox provided. I'm going to see if I can figure out what is going on

**Comment by bluebill1049** [comment]
@ASHFAQPATWARI thanks for the clarification on the key usage.

Thanks @stramel I will investigate it bit deeper, it related to `ref` getting outdated or switched due to `key` change. I suspect its a React or like you said how material UI handle `ref`. I will let you guys know after my further investigation.

**Comment by bluebill1049** [comment]
It is not `material-UI` lead to this issue, and not caused by this lib manipulate `ref`.

I have removed `key` from your example and the style still works fine for me, I am not clear why `key` is needed to recalculated style. Keys in React are utilized to identify specific Virtual DOM Elements that have changed. Would be helpful to point me to the docs for the usage. but by changing the `key` definitely confused `react` on the `ref` which lead to this weird behavior.

@stramel FYI, if you have time please help verify my investigation. 👍

**Comment by stramel** [comment]
@bluebill1049 When you remove the `key` the input border doesn't update to the width of the label when the language changes. I'm not sure if there is a better way to handle it though. Also, I can't find anything on the documentation stating that `key` is how to handle it. @ASHFAQPATWARI Can you point us to that documentation?

**Comment by bluebill1049** [comment]
@stramel it seems very strange to change a key on a component. This is the first time that I have seen such a usage, maybe I miss the big picture here (material-ui). I only had to use `key` for `children` (during `map`). I am pretty sure this is the reason which confused the react to believe it's a different component and hence `ref` got removed (while the second render secured the ref again) Yea I would like to see the documentation on material UI for key usage as well and find out the reason behind.

**Comment by ASHFAQPATWARI** [comment]
@bluebill1049 Using key prop is a workaround suggested by Material-ui Team here: https://github.com/mui-org/material-ui/issues/15142#issuecomment-478583219

**Comment by bluebill1049** [comment]
> One workaround is to change the key on the text field when the label width change

It's a workaround, and I agree with ith that. Is there any way you could toggle the UI with state pass down instead of changing the `key`?

**Comment by ASHFAQPATWARI** [comment]
Without changing the `key`, the inline styles are not being recalculated. I am already toggling the UI by changing the context value.

**Comment by bluebill1049** [comment]
That's unfortunate. I have to say this is such an edge case. There is nothing I can do about this issue at this lib level (i have already spent many hrs on this) because of React is confused when u switch the `key` and lead `ref` to be `undefined` which is not standard anyway. the only solution I can think about it is custom register, which means not involved with `ref`

```
useEffect(() => register({ name: xxx }), [ register ])
```

https://react-hook-form.com/api

you can read more on the custom register on the above link too.

**Comment by leapful** [comment]
@bluebill1049 I have the same issue while using Material-UI with react-form-hook. After debugging, I found that the issue is a side effect from method findRemovedFieldAndRemoveListener().

For example, in sample codesandbox of @ASHFAQPATWARI, when switching language, Login component will be re-rendered and method registerIntoFieldRef will be triggered too.'

https://github.com/bluebill1049/react-hook-form/blob/88e29b696f265ffdafd329e038dc0842f75a9dd2/src/useForm.ts#L379

But at this time, fieldsRef already has value (referenced to old HTML Input element of email and password) therefore fieldsRef is not updated.

Right after the execution of registerIntoFieldRef(), JSS library (used by Material-UI) is executed to update style for input component but it also triggers callback on MutationObserver from method onDomRemove(). 

https://github.com/bluebill1049/react-hook-form/blob/681fd2f824f116dd749075ce42d01f04a7e009b5/src/utils/onDomRemove.ts#L3

Because the HTML Input element from parameter of onDomRemove() is the old element that does not exist in the current rendering page, method isDetached() returns true and onDetachCallback() will be triggered and this callback is findRemovedFieldAndRemoveListener().

https://github.com/bluebill1049/react-hook-form/blob/681fd2f824f116dd749075ce42d01f04a7e009b5/src/utils/onDomRemove.ts#L19

At this time, findRemovedFieldAndRemoveListener() will remove all values in fieldsRef.

https://github.com/bluebill1049/react-hook-form/blob/681fd2f824f116dd749075ce42d01f04a7e009b5/src/logic/findRemovedFieldAndRemoveListener.ts#L5

Now all values in fieldsRef has been removed, when clicking on submit button, method handleSubmit() always gets empty values from method getFieldsValues() and causes the wrong validation.

https://github.com/bluebill1049/react-hook-form/blob/681fd2f824f116dd749075ce42d01f04a7e009b5/src/useForm.ts#L537

In summary, the root cause is from findRemovedFieldAndRemoveListener() which is executed on old data by MutationObserver on document when style is changed.

A workaround is to put unSubscribe() inside Login component to clean up old data before re-rendering.

Updated sample:
https://codesandbox.io/s/react-hook-form-issue-on-first-click-ffhr0

**Comment by bluebill1049** [comment]
wow thanks, @leapful for the detailed investigation! I would love to get this working as well. but a toggling individual element with the key is definitely not standard (edge case). I will be looking into this issue see if I can work it out. Feel free to send a PR if you work out a solution too. much appreciated your investigation.

**Comment by ASHFAQPATWARI** [comment]
@leapful Thats some really inspiring investigation.

**Comment by bluebill1049** [comment]
Ok here is a solution for "hacking" `key` value with `material-ui`

**Problem**: (thanks @leapful based on your investigation)

- toggle `key` value on the same component
- confused React to remove the existing Element and then insert a new one (due to tree diffing [virtual DOM])
- confused React hook form because the name is the same however the actual dom is removed.
- when the element gets removed React hook form will remove all event listeners associated with the element to avoid memory leak
- re-render happened from context (provider), `register` gets invoked with old `ref` then React work out tree diff, remove the old node and insert the new one. However `register` is always called before that gets happened not after, hence `registered` ref lead to be the deleted one and empty.

**Solution** 

- called hidden API `unsubscribe` on each render, which was suggested by @leapful 
it works, but terrible from the performance point of view, also his hidden API have been removed in the newest version.

- manually trigger a re-render after context update to make sure `reigster` get the correct ref after React reconciliation with the diffing, remove and insert.
https://codesandbox.io/s/react-hook-form-issue-on-first-click-uxvhx

ideally, `material-ui` should fix the bug on the context API, which stops users from toggling `key` on the same component to working around the re-render issue.

**Comment by leapful** [comment]
@bluebill1049 Nice trick to update ref :+1: Actually, we can just use "lang" of Login component props as dependency for useEffect.

```javascript
const Login = ({lang}) => {
    ...
    const [, forceUpdate] = React.useState({});

    React.useEffect(() => forceUpdate({}), [lang]);
    ...
};
```

**Comment by bluebill1049** [comment]
i think we can close this issue now @ASHFAQPATWARI hope material UI will sort this bug out soon, so we dont have to do this trick

**Comment by bluebill1049** [comment]
i am closing this issue, feel free to re-open for more questions.

**Comment by bluebill1049** [comment]
If you are facing similar issue, material-ui is considering to support react-hook-form. https://github.com/react-hook-form/react-hook-form/issues/477
